### PR TITLE
Add libsvm format design doc

### PR DESCRIPTION
**Pull request by sneaxiy**
Support LibSVM format in SQLFlow. Related to https://github.com/sql-machine-learning/sqlflow/issues/2323 .

### CODE REVIEW

File: doc/design/libsvm_format_column.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# Support LibSVM format on SQLFlow
+
+## Background
+
+SQLFlow extends SQL syntax to support End-to-End machine learning pipeline.
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: End-to-End => end-to-end.  Use grammarly.com pleaase.

File: doc/design/libsvm_format_column.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
+# Support LibSVM format on SQLFlow
+
+## Background
+
+SQLFlow extends SQL syntax to support End-to-End machine learning pipeline. 
+However, due to the variety of the data format (like CSV, LibSVM, etc.),
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: libsvm is not a file format. According to https://www.csie.ntu.edu.tw/~cjlin/libsvm/, LIBSVM is "A Library for Support Vector Machines". What do you really want to say here?

File: doc/design/libsvm_format_column.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
+# Support LibSVM format on SQLFlow
+
+## Background
+
+SQLFlow extends SQL syntax to support End-to-End machine learning pipeline. 
+However, due to the variety of the data format (like CSV, LibSVM, etc.), 
+data pre-processing is usually a tough task for users. Users have to speed lots of time
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: speed => spend?

Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: It is not the user who transforms the raw inputs.  It is the program that transforms the raw inputs.

File: doc/design/libsvm_format_column.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+## Background
+
+SQLFlow extends SQL syntax to support End-to-End machine learning pipeline. 
+However, due to the variety of the data format (like CSV, LibSVM, etc.), 
+data pre-processing is usually a tough task for users. Users have to speed lots of time 
+to transform the raw inputs into the data format which the machine learning model accepts.
+SQLFlow should do some automatic data pre-processing to make End-to-End machine learning pipeline 
+easier.
+
+LibSVM is a commonly used data format in machine learning, and it is officially supported in many machine learning frameworks,
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: Again, LIBSVM is not a data format. The authors of LIBSVM defined a data [file format](https://www.csie.ntu.edu.tw/~cjlin/libsvm/faq.html#f307) -- remember to provide the URL to the format definition.

File: doc/design/libsvm_format_column.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+SQLFlow extends SQL syntax to support End-to-End machine learning pipeline. 
+However, due to the variety of the data format (like CSV, LibSVM, etc.), 
+data pre-processing is usually a tough task for users. Users have to speed lots of time 
+to transform the raw inputs into the data format which the machine learning model accepts.
+SQLFlow should do some automatic data pre-processing to make End-to-End machine learning pipeline 
+easier.
+
+LibSVM is a commonly used data format in machine learning, and it is officially supported in many machine learning frameworks, 
+such as [XGBoost](https://xgboost.readthedocs.io/en/latest/tutorials/input_format.html) and
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: XGBoost => XGBoost input data file format

File: doc/design/libsvm_format_column.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+SQLFlow should do some automatic data pre-processing to make End-to-End machine learning pipeline 
+easier.
+
+LibSVM is a commonly used data format in machine learning, and it is officially supported in many machine learning frameworks, 
+such as [XGBoost](https://xgboost.readthedocs.io/en/latest/tutorials/input_format.html) and 
+[SkLearn](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_svmlight_file.html). There have been 
+[some issues](https://github.com/sql-machine-learning/sqlflow/issues/2323) that require SQLFlow should support LibSVM format column. 
+Therefore, SQLFlow should support to process LibSVM format to simplify the End-to-End machine learning pipeline.
+
+LibSVM format is more portable than CSV format when storing sparse data. Each line of the LibSVM file is like:
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: LibSVM format => LibSVM data file format

File: doc/design/libsvm_format_column.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+```
+
+where:
+
+- `<label>` is optional (for example, `<label>` is optional when predicting). 
+- `<index>:<value>` indicates each non-empty index-value pair in the sparse data. There may be many index-value pairs in each line, 
+and the pair number may be different in different lines.
+
+
+There are 2 ways in SQLFlow to support the column which is in LibSVM format:
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: What is the relationship between column and the data file format?  The readers, including me, would feel abrupt to see the new concept column -- what is it?

File: doc/design/libsvm_format_column.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+If we want to support LibSVM data format, where would be 2 optional methods:
+    - Add a new `COLUMN` clause, like `LIBSVM` or `NUMERIC_KV`. It introduces complexity, and disobeys Occam's Razor principle.
+    - Extend `NUMERIC` column to support LibSVM format. Currently, we have supported CSV format implicitly: in feature derivation stage, 
+    we would infer whether the column data is CSV format (see [here](https://github.com/sql-machine-learning/sqlflow/blob/3b70a0599beef573cd99f15dd41cc0a194634b75/pkg/ir/derivation.go#L146)),
+    and convert the CSV format value into Tensor in Python side (see [here](https://github.com/sql-machine-learning/sqlflow/blob/develop/python/sqlflow_submitter/db.py#L159)). 
+    We can do the same implicit feature derivation to LibSVM format as CSV format.
+    
+In conclusion, we would choose to extend `NUMERIC` column to support LibSVM format column in this design.
+
+## Only support LibSVM data without label value
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: Section header must be capitalized.

File: doc/design/libsvm_format_column.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            }
+        }
+        rowCount ++
+    }
+}
+```
+
+### Convert the LibSVM data to sparse Tensor in Python data generator
+
+Whether the column is in LibSVM format is inferred in feature derivation stage, and the data format information would be
```
Comment (wangkuiyi) [inline doc/design/libsvm_format_column.md]: No passive voice. Always use active voice in technical writings.  Grammarly.com would remind you about the use of passive voice, so definitely use Grammarly.com anytime before submitting English writings. I would suggest you install the Chrome extension of Grammarly.com.

File: doc/design/libsvm_format_column.md (hunk 1 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Done.

File: doc/design/libsvm_format_column.md (hunk 2 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: I have added the `Definition` section to make it more clear. Please see [here](https://github.com/sql-machine-learning/sqlflow/pull/2554/files#diff-f73c2f4f93fc028434d604606630b34dR3).

File: doc/design/libsvm_format_column.md (hunk 3 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: This sentence has been removed after polishing.

Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: This sentence has been removed after polishing.

File: doc/design/libsvm_format_column.md (hunk 4 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: I have added the `Definition` section to make it more clear. Please see [here](https://github.com/sql-machine-learning/sqlflow/pull/2554/files#diff-f73c2f4f93fc028434d604606630b34dR3).

File: doc/design/libsvm_format_column.md (hunk 5 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: This sentence is changed to be: `Spark MLLib, XGBoost and SkLearn officially support loading data from this kind of file for training and prediction.` Please see [here](https://github.com/sql-machine-learning/sqlflow/pull/2554/files#diff-f73c2f4f93fc028434d604606630b34dR23).

File: doc/design/libsvm_format_column.md (hunk 8 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Done.

File: doc/design/libsvm_format_column.md (hunk 9 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Done.

File: doc/design/libsvm_format_column.md (hunk 7 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: In the new-added `Definition` section, I have defined 2 concepts:

- LibSVM format file: text file whose each line is in the format like `<label> <index>:<value> <index>:<value>...`.
- LibSVM format column: DBMS column, whose value of each row is in the same format as each line of a LibSVM format file.

File: doc/design/libsvm_format_column.md (hunk 6 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Added the "LibSVM format file" definition at the beginning of the design doc.

File: doc/design/libsvm_format_column.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
+# Support LibSVM Format Column in SQLFlow
+
+## Definition
+
+### LibSVM Format File
```
Comment (typhoonzero) [inline doc/design/libsvm_format_column.md]: Can put this to the last section "Appendix".

File: doc/design/libsvm_format_column.md [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+A **LibSVM format column** is a DBMS table column, whose value of each row is in the same format as each line in a **LibSVM format file**.
+
+For example, in the following table, `c1` is a LibSVM format column, while `c2` is not a LibSVM format column.
+
+```
++-------------------- +------+
+| c1                  | c2   |
++---------------------+------+
+| 0 1:2.1 3:4.5 5:6.0 |  2.8 |
```
Comment (typhoonzero) [inline doc/design/libsvm_format_column.md]: If this represents the example, remove the label here.

File: doc/design/libsvm_format_column.md [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+- `COLUMN` clauses. The `COLUMN` clauses in SQLFlow support some commonly used feature columns, like `NUMERIC`, `CATEGORY_ID`, `EMBEDDING`, etc. If we want to support loading the data from the LibSVM format column, where would be 2 optional methods:
+    - Add a new `COLUMN` clause, like `LIBSVM` or `NUMERIC_KV`. It introduces complexity and disobeys the Occam's Razor principle.
+    - Extend the `NUMERIC` column clause to support columns in different data formats. Currently, we have supported CSV format implicitly in `NUMERIC` column clause: in feature derivation stage, we would infer whether the column data is in CSV format (see [here](https://github.com/sql-machine-learning/sqlflow/blob/3b70a0599beef573cd99f15dd41cc0a194634b75/pkg/ir/derivation.go#L146)), and convert the values in CSV format into tensors in Python side (see [here](https://github.com/sql-machine-learning/sqlflow/blob/develop/python/sqlflow_submitter/db.py#L159)). We can do the same implicit feature derivation and conversion for the LibSVM format column.
+    
+In conclusion, we would choose to extend the `NUMERIC` column to support the LibSVM format column in this design.
```
Comment (typhoonzero) [inline doc/design/libsvm_format_column.md]: Use `SPARSE` should be better, since we already describe sparse inputs using `SPARSE` syntax in column clause. e.g. `COLUMN NUMERIC(SPARSE(col, [10000], "\s"), shape)`

**Review by typhoonzero — DISMISSED** [review]

File: doc/design/libsvm_format_column.md (hunk 12 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: There may be some problems if we use `SPARSE` :

- Both `NUMERIC` and `SPARSE` have the same parameter `shape`. There may be a conflict.
- I have tested that TensorFlow does not support `SparseTensor` as the feature input of `tf.feature_column.numeric_column`. `SparseTensor` would only be used in the `EMBEDDING`, `CATEGORY_ID` layer.

I think that we should reconsider whether `SPARSE` should be used together with the `NUMERIC` column.

File: doc/design/libsvm_format_column.md (hunk 11 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Done.

File: doc/design/libsvm_format_column.md (hunk 10 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: I prefer to put this into the first section to make readers more clear about the important concepts in the design doc.

File: doc/design/libsvm_format_column.md [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+For example, the text line `0 1:2 3:4 5:6` in a LibSVM format file indicates that:
+
+- there is a sparse sample `x`, and `x[1]=2`, `x[3]=4`, `x[5]=6`, and the other values of sample `x` are zeros.
+- the label corresponding to the sample is 0.
+
+This kind of file was firstly introduced in [LibSVM](https://www.csie.ntu.edu.tw/~cjlin/libsvm), a library for Support Vector Machines algorithms. After that, this kind of file is widely used in many other machine learning frameworks. For example, [Spark MLLib](https://spark.apache.org/docs/1.0.2/api/python/pyspark.mllib.util.MLUtils-class.html), [XGBoost](https://xgboost.readthedocs.io/en/latest/tutorials/input_format.html) and [SkLearn](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_svmlight_file.html) officially support loading data from this kind of file for training and prediction.
+
+We call this kind of file to be **LibSVM format file** in this document, as all of the machine learning frameworks mentioned above call this file format to be **LibSVM format**.
+
+### LibSVM Format Column
```
Comment (brightcoder01) [inline doc/design/libsvm_format_column.md]: The table is not exactly the same with LibSVM format. In the example, the label is in a separate column. For LibSVM, the label and KV features are stored in one column in the table.

File: doc/design/libsvm_format_column.md [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            }
+        }
+        rowCount ++
+    }
+}
+```
+
+### Convert the Values of the LibSVM Format Column to Sparse Tensor in Python Side
+
+After we have inferred the format of each column in the feature derivation stage, and the data format information would be used in Python code generation. In the Python side, we would first read the raw data from the database, and then convert the data from the LibSVM format column into sparse tensors for training, prediction, evaluating, and explaining.
```
Comment (brightcoder01) [inline doc/design/libsvm_format_column.md]: For XGBOOST，we won’t convert the data to tensor

Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Thanks! Fixed.

File: doc/design/libsvm_format_column.md (hunk 13 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Yes. The standard LibSVM format would contain the label, but it is not required. For example, XGBoost does support to load LibSVM format without the label, and XGBoost calls this kind of file to be LibSVM format as well. In this design, we only support the LibSVM format without the label. Please see [here](https://github.com/sql-machine-learning/sqlflow/pull/2554/files#diff-f73c2f4f93fc028434d604606630b34dR57).

Comment (wangkuiyi) [reply doc/design/libsvm_format_column.md]: So it is not anything like "LibSVM format". We are not supposed to mention LibSVM in this document.

File: doc/design/libsvm_format_column.md (hunk 10 above)
Comment (wangkuiyi) [reply doc/design/libsvm_format_column.md]: It doesn't help readers to understand the document. It is confusing. SQLFlow doesn't read files in the LibSVM format at all.

File: doc/design/libsvm_format_column.md [hunk 15]
```diff
@@ -0,0 +1,146 @@
+# Load Sparse Tensors From LibSVM Format Column by SPARSE Column Clause in SQLFlow
```
Comment (brightcoder01) [inline doc/design/libsvm_format_column.md]: I think XGBoost don’t have the tensor concept

File: doc/design/libsvm_format_column.md [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+```
++-------------------+------+
+| c1                | c2   |
++-------------------+------+
+| 1:2.1 3:4.5 5:6.0 |  2.8 |
+| 7:-3.2 9:2.3      |  3.1 |
++-------------------+------+
+```
+
+## Background
```
Comment (brightcoder01) [inline doc/design/libsvm_format_column.md]: IMO, background is usually the first section

File: doc/design/libsvm_format_column.md [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    - If `"3,5,7"` represents a sparse vector, each value of the sparse tensor `x` would only be 0 or 1, and `x[3] = x[5] = x[7] = 1`. 
+
+We should support loading data from the LibSVM format column in SQLFlow, because:
+
+- The LibSVM format column is a common way to store the sparse data in DBMS.
+- There is [an issue](https://github.com/sql-machine-learning/sqlflow/issues/2323) that requires SQLFlow should support loading data from the LibSVM format column.
+
+There would be 2 ways to load data from the LibSVM format column in SQLFlow:
+
+- `TO RUN` statements. It can support any user-defined data pre-processing. We can release a docker image to convert the data from the LibSVM format column to a dense tensor, create columns for each element of the dense tensor, and write the data into the result table. However, if the data is highly sparse, it would create too many columns in the result table, and consume lots of time and memories to do the transformation.
```
Comment (brightcoder01) [inline doc/design/libsvm_format_column.md]: We can add the design doc link here

File: doc/design/libsvm_format_column.md (hunk 15 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Changed `tensor` to `vector`.

File: doc/design/libsvm_format_column.md (hunk 16 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: Removed the `Definitions` section.

File: doc/design/libsvm_format_column.md (hunk 17 above)
Comment (sneaxiy) [reply doc/design/libsvm_format_column.md]: In the new polished doc, the explanation of `TO RUN` is removed.

**Review by wangkuiyi — APPROVED** [review]
